### PR TITLE
Add ingress/egress info to network-get

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

The network-get hook tool now prints ingress and egress information about the relation 
The --primary-address arg is deprecated in favour of 3 new args:
--ingress-address (the address the local unit should advertise for incoming connections)
--binding-address (the address the local unit should use to listen on)
--egress-subnets (subnets from which outgoing traffic from the local unit originates)

Without args, a full yaml data structure is printed, eg
`$ network-get db -r 1`
`info:`
`- macaddress: "00:11:22:33:44:33"`
`  interfacename: eth3`
`  addresses:`
`  - address: 10.33.1.8`
`    cidr: 10.33.1.8/24`
`egress-subnets:`
`- 192.168.1.0/8`
`- 10.0.0.0/8`
`ingress-addresses:`
`- 100.1.2.3`
`- 100.4.3.2`

Note that for backwards compatibility, binding addresses are in the info block, which also contains other link layer data like MAC address etc.

Similar to how --primary-address behaves, if --ingress-address or --binding-address is specified, a single value is printed.

$ network-get db -r 1 --ingress-address
100.1.2.3

$ network-get db -r 1 --binding-address
10.33.1.8

## QA steps

Run up a cmr scenario in aws; mysql in one model, mediawiki in another.
juju run network-get to check the output
- in cmr relation context -> ingress-address is public address of machine
- no relation context or non cmr relation -> ingress address is private address of machine
- set egress subnets, check these are printed when present

### CODE REVIEW

File: apiserver/facades/agent/uniter/uniter.go [hunk 1]
```diff
@@ -1837,37 +1847,57 @@ func (u *UniterAPI) NetworkInfo(args params.NetworkInfoParams) (params.NetworkIn
...
-		}
-	}
 	for binding, space := range bindingsToSpace {
-		result.Results[binding] = networkingcommon.MachineNetworkInfoResultToNetworkInfoResult(networkInfos[space])
+		// The binding address information based on link layer devices.
+		info := networkingcommon.MachineNetworkInfoResultToNetworkInfoResult(networkInfos[space])
+
+		// Set egress and ingress address information.
+		info.EgressSubnets = bindingsToEgressSubnets[binding]
+		// If there is no ingress address explicitly defined for a given binding,
```
Comment (axw) [inline apiserver/facades/agent/uniter/uniter.go]: I don't get this. If we have room for multiple addresses, why include only one? Charms can just pick the first one if they want to.

Comment (wallyworld) [reply apiserver/facades/agent/uniter/uniter.go]: This was a carry over from older code. We now return all the addresses.

Comment (wallyworld) [reply apiserver/facades/agent/uniter/uniter.go]: It now adds all addresses into the ingress address slice

File: apiserver/params/network.go [hunk 2]
```diff
@@ -688,8 +688,10 @@ type NetworkInfo struct {
 
 // NetworkInfoResult holds either and error or a list of NetworkInfos for given binding.
 type NetworkInfoResult struct {
-	Error *Error        `json:"error,omitempty" yaml:"error,omitempty"`
-	Info  []NetworkInfo `json:"network-info" yaml:"info"`
+	Error            *Error        `json:"error,omitempty" yaml:"error,omitempty"`
+	Info             []NetworkInfo `json:"bind-addresses" yaml:"bind-addresses"`
```
Comment (axw) [inline apiserver/params/network.go]: this should be omitempty, in case error != nil

File: worker/uniter/runner/jujuc/network-get.go [hunk 3]
```diff
@@ -16,26 +16,46 @@ type NetworkGetCommand struct {
...
 	doc := `
 network-get returns the network config for a given binding name. By default
 it returns the list of interfaces and associated addresses in the space for
-the binding.
-If --primary-address flag is specified then only single IP address is
-returned that the local unit should advertise as its endpoint to its peers.
+the binding, as well as the ingress address for the binding. If defined, any
+egress subnets are also returned.
+If one of the following flags are specified, just that value is returned.
+If more then one flag is specified, a map of values is returned.
```
Comment (axw) [inline worker/uniter/runner/jujuc/network-get.go]: s/then/than/

File: worker/uniter/runner/jujuc/network-get.go [hunk 4]
```diff
@@ -78,12 +118,43 @@ func (c *NetworkGetCommand) Run(ctx *cmd.Context) error {
...
 
+	// If no specific attributes asked for,
+	// print everything we know.
+	if !c.primaryAddress && len(c.keys) == 0 {
+		return c.out.Write(ctx, ni)
+	}
+
+	// Backwards compatibility - we just want the primary address.
 	if c.primaryAddress {
 		if len(ni.Info[0].Addresses) == 0 {
```
Comment (axw) [inline worker/uniter/runner/jujuc/network-get.go]: error if the user specified any other flags?

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**